cpp/include/raft/error.hpp

```diff
@@ -0,0 +1,159 @@
+  explicit exception(std::string const msg) noexcept
+    : std::exception(), msg_(std::move(msg)) {
+    collect_call_stack();
+  }
+
+  /** get the message associated with this exception */
+  char const* what() const noexcept override { return msg_.c_str(); }
+
+ private:
+  /** message associated with this exception */
+  std::string msg_;
+
+  /** append call stack info to this exception's message for ease of debug */
+  // Courtesy: https://www.gnu.org/software/libc/manual/html_node/Backtraces.html
+  void collect_call_stack() noexcept {
+#ifdef __GNUC__
+    constexpr int kMaxStackDepth = 64;
+    void* stack[kMaxStackDepth];  // NOLINT
+    auto depth = backtrace(stack, kMaxStackDepth);
+    std::ostringstream oss;
+    oss << std::endl << "Obtained " << depth << " stack frames" << std::endl;
+    char** strings = backtrace_symbols(stack, depth);
+    if (strings == nullptr) {
+      oss << "But no stack trace could be found!" << std::endl;
+      msg_ += oss.str();
+      return;
+    }
+    ///@todo: support for demangling of C++ symbol names
+    for (int i = 0; i < depth; ++i) {
+      oss << "#" << i << " in " << strings[i] << std::endl;
+    }
+    free(strings);
+    msg_ += oss.str();
+#endif  // __GNUC__
+  }
+};
+
+/**
+ * @brief Exception thrown when logical precondition is violated.
+ *
+ * This exception should not be thrown directly and is instead thrown by the
+ * RAFT_EXPECTS and  RAFT_FAIL macros.
+ *
+ */
+struct logic_error : public raft::exception {
+  explicit logic_error(char const* const message) : raft::exception(message) {}
+  explicit logic_error(std::string const& message) : raft::exception(message) {}
+};
+
+}  // namespace raft
+
+// FIXME: Need to be replaced with RAFT_FAIL
+/** macro to throw a runtime error */
+#define THROW(fmt, ...)                                                        \
+  do {                                                                         \
+    std::string msg;                                                           \
+    char errMsg[2048]; /* NOLINT */                                            \
+    std::snprintf(errMsg, sizeof(errMsg),                                      \
+                  "exception occured! file=%s line=%d: ", __FILE__, __LINE__); \
+    msg += errMsg;                                                             \
+    std::snprintf(errMsg, sizeof(errMsg), fmt, ##__VA_ARGS__);                 \
+    msg += errMsg;                                                             \
+    throw raft::exception(msg);                                                \
+  } while (0)
+
+// FIXME: Need to be replaced with RAFT_EXPECTS
+/** macro to check for a conditional and assert on failure */
+#define ASSERT(check, fmt, ...)              \
+  do {                                       \
+    if (!(check)) THROW(fmt, ##__VA_ARGS__); \
+  } while (0)
+
+#define SET_ERROR_MSG(msg, location_prefix, fmt, ...)                        \
```

Comment:
minor nitpick: `location_prefix` seems to be unused?

Reply:
I guess it was supposed to be used in the first `snprintf` call below?

Reply:
Nice catch!!!